source/code/plugins/agent_maintenance_script.rb

```diff
@@ -0,0 +1,639 @@
+      rescue => e
+        error = e
+      ensure
+        key_file.close
+        cert_file.close
+      end
+
+      # Check for any error or non-existent or empty files
+      if !error.nil? or !file_exists_nonempty(@cert_path) or !file_exists_nonempty(@key_path)
+        log_error("Error generating certs")
+        return 1
+      end
+
+      return 0
+    end
+
+    # Simple class to support interaction with topology script helper method (obj_to_hash)
+    class AgentRenewCertificateRequest < StrongTypedClass
+      strongtyped_accessor :NewCertificate, String
+    end
+
+    # Restore the provided public/private key to the certs files
+    def restore_old_certs(cert_old, key_old)
+      cert_file = File.open(@cert_path, "w")
+      cert_file.write(cert_old)
+      cert_file.close
+
+      key_file = File.open(@key_path, "w")
+      key_file.write(key_old)
+      key_file.close
+    end
+ 
+    # Renew certificates for agent/workspace connection
+    def renew_certs
+      # Check necessary inputs
+      if @load_config_success != 0
+        log_error("Error loading configuration from #{@omsadmin_conf_path}")
+        return 1
+      elsif @WORKSPACE_ID.nil? or @AGENT_GUID.nil? or @WORKSPACE_ID.empty? or @AGENT_GUID.empty?
+        log_error("Missing required field from configuration file: #{@omsadmin_conf_path}")
+        return 1
+      elsif @CERTIFICATE_UPDATE_ENDPOINT.nil? or @CERTIFICATE_UPDATE_ENDPOINT.empty?
+        log_error("Missing CERTIFICATE_UPDATE_ENDPOINT from configuration")
+        return 1
+      elsif !file_exists_nonempty(@cert_path) or !file_exists_nonempty(@key_path)
+        log_error("No certificates exist; cannot renew certificates")
+        return 1
+      end
+
+      log_info("Renewing the certificates")
+
+      # Save old certs
+      cert_old = OpenSSL::X509::Certificate.new(File.open(@cert_path))
+      key_old = OpenSSL::PKey::RSA.new(File.open(@key_path))
+
+      generated = generate_certs(@WORKSPACE_ID, @AGENT_GUID)
+      if generated != 0
+        return 1
+      end
+
+      # Form POST request
+      renew_certs_req = AgentRenewCertificateRequest.new
+      renew_certs_req.NewCertificate = get_cert_server(@cert_path)
+
+      renew_certs_xml = "<?xml version=\"1.0\"?>\n"
+      renew_certs_xml.concat(Gyoku.xml({ "CertificateUpdateRequest" => {:content! => obj_to_hash(renew_certs_req), \
+:'@xmlns:xsi' => "http://www.w3.org/2001/XMLSchema-instance", :'@xmlns:xsd' => "http://www.w3.org/2001/XMLSchema", \
+:@xmlns => "http://schemas.microsoft.com/WorkloadMonitoring/HealthServiceProtocol/2014/09/"}}))
+
+      uri = URI.parse(@CERTIFICATE_UPDATE_ENDPOINT)
+      req = Net::HTTP::Post.new(uri.request_uri, headers = {})
+      req.body = renew_certs_xml
+
+      # Form HTTP
+      http = OMS::Common.create_secure_http(uri, get_proxy_info)
+      http.cert = cert_old
+      http.key = key_old
+
+      print("Generated renew certificates request:\n#{req.body}\n") if @verbose
```

Comment:
I used print instead of log for the verbose outputs in an effort to mimic omsadmin's current behavior. However, I do agree they should be logged.